### PR TITLE
[sender] enable buffer pool + concurrent sending threads

### PR DESCRIPTION
**Pull request by truthbk**
This PR adds a buffer pool to the sender. The original code used a single pre-allocated buffer we reused to submit batched dogstatsd metrics. While efficient from a memory perspective, this approach implied we had to stop processing dogstatsd metrics while waiting on IO for every packet submission. 

This PR aims to decouple actual packet transmission from dogstatsd message processing and packet assembling. To achieve this we introduce a pool of preallocated buffers and an additional queue, message are processed and batched into these buffers - blocking, if no buffers are available (further improvements are possible here, where we could entertain different strategies to deal with buffer unavailability), once a packet has been filled or no more messages are waiting, the buffer is queued in an outgoing queue. A workers thread (or threads) then will dequeue and write these buffers to the socket. The benefits of this approach, other than the high-level idea explained above, is that it should allow for better usage of the IOwait slice, as the OS kernel will likely do the right thing when scheduling pending work helping us increase the throughput.

TL;DR:
- decouple processing of incoming dogstatsd messages from packet submission.
- worker thread(s) for packet submission should allow the OS to better schedule work in the active client threads.
- finer granular control with new knobs (queue size, number of sender threads, etc).

Note: the max perf test is useful for benchmarking but might be worth removing or modifying before merging - or disabling in the CI. It also conflicts currently with a similar test in `jaime/perf`, we can address that later.

Note: this should be merged onto `jaime/perf`  (so before #94)

### CODE REVIEW

File: src/main/java/com/timgroup/statsd/StatsDProcessor.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        this.messages = new ConcurrentLinkedQueue<String>();
+        this.outboundQueue = new ArrayBlockingQueue<ByteBuffer>(poolSize);
+
+        this.handler = handler;
+    }
+
+    boolean send(final String message) {
+        if (!shutdown) {
+            if (qSize.get() < qCapacity) {
+                messages.offer(message);
```
Comment (ogaca-dd) [inline src/main/java/com/timgroup/statsd/StatsDProcessor.java]: Returned value of `boolean java.util.Queue.offer(String e)`  might be checked.

File: src/main/java/com/timgroup/statsd/StatsDProcessor.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+                    }
+                    sendBuffer.put(data);
+                    if (null == messages.peek()) {
+                        outboundQueue.put(sendBuffer);
+                        sendBuffer = bufferPool.borrow();
+                    }
+                }
+            } catch (final InterruptedException e) {
+                if (shutdown) {
+                    return;
```
Comment (ogaca-dd) [inline src/main/java/com/timgroup/statsd/StatsDProcessor.java]: Do we also want to call `handler.handle(e);` as in line 76?

File: src/main/java/com/timgroup/statsd/BufferPool.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    ByteBuffer borrow() throws InterruptedException {
+        return pool.take();
+    }
+
+    void put(ByteBuffer buffer) throws InterruptedException {
+        pool.put(buffer);
+    }
+
+    int available() {
```
[the remark this answers is not in the document]
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/BufferPool.java]: I still like it as a class method, so I think I'm going to keep it unless you strongly disagree, good catch.

File: src/main/java/com/timgroup/statsd/StatsDProcessor.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
+package com.timgroup.statsd;
+
+import java.io.IOException;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.DatagramChannel;
+import java.nio.charset.Charset;
+import java.util.concurrent.Callable;
+
```
[the remark this answers is not in the document]
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDProcessor.java]: 👌

File: src/main/java/com/timgroup/statsd/StatsDProcessor.java (hunk 2 above)
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDProcessor.java]: It's a good question, I'm not sure, we're going to gracefully shutdown, and handle doesn't do much, but it's a good question. 🤔

Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDProcessor.java]: Let me think about it, I'll address in #94 if I decide we should handle these errors differently.

File: src/main/java/com/timgroup/statsd/StatsDProcessor.java (hunk 1 above)
Comment (truthbk) [reply src/main/java/com/timgroup/statsd/StatsDProcessor.java]: This has now been moved to `StatsDNonBlockingProcessor`, there we use a `ConcurrentLinkedQueue` which never returns `false` on `offer()` because it is an unbounded queue.

**Comment by truthbk** [comment]
CI is green, merging to the big larger #94, final changes and fixes will be made there. 

Thank you for the review.
